api/api/versions_api.go

```diff
@@ -145,6 +146,108 @@ func (c *VersionsController) CreateVersion(r *http.Request, vars map[string]stri
+	}
+
+	// handle for pyfunc v2, since the prediction job feature is only available for pyfunc_v2 model
+	if model.Type == "pyfunc_v2" {
+		// check active prediction job
+		// if there are any active prediction job using the model version, deletion of the model version are prohibited
+		httpStatus, jobs, err := c.checkActivePredictionJobs(ctx, model, version)
+		if err != nil {
+			return NewError(httpStatus, err.Error())
+		}
+
+		// delete inactive prediction job
+		for _, item := range jobs {
+			_, err = c.PredictionJobService.StopPredictionJob(ctx, item.Environment, model, version, item.ID)
+			if err != nil {
+				log.Errorf("failed to stop prediction job %v", err)
+				return InternalServerError(fmt.Sprintf("Failed stopping prediction job: %s", err))
+			}
+		}
+
+	} else {
+		// handle for model with type non pyfunc
+		// check active endpoints
+		// if there are any active endpoint using the model version, deletion of the model version are prohibited
+		httpStatus, endpoints, err := c.checkActiveEndpoints(ctx, model, version)
+		if err != nil {
+			return NewError(httpStatus, err.Error())
+		}
+		// delete inactive endpoint
+		for _, item := range endpoints {
+			err = c.EndpointsService.DeleteEndpoint(version, item)
+			if err != nil {
+				log.Errorf("failed to undeploy endpoint job %v", err)
+				return InternalServerError(fmt.Sprintf("Failed to delete endpoint: %s", err))
+			}
+		}
+	}
+	// delete mlflow run and artifact
+	err = c.MlflowDeleteService.DeleteRun(ctx, version.RunID, version.ArtifactURI, true)
+	if err != nil {
+		log.Errorf("failed to delete mlflow run %v", err)
+		return InternalServerError(fmt.Sprintf("Delete Failed: %s", err.Error()))
+	}
+
+	// delete model version from db
+	err = c.VersionsService.Delete(version)
+	if err != nil {
+		log.Errorf("failed to delete model version %v", err)
+		return InternalServerError(fmt.Sprintf("Delete Failed: %s", err.Error()))
+	}
+
+	return Ok(versionID)
+}
+func (c *VersionsController) checkActivePredictionJobs(ctx context.Context, model *models.Model, version *models.Version) (int, []*models.PredictionJob, error) {
+	jobQuery := &service.ListPredictionJobQuery{
+		ModelID:   model.ID,
+		VersionID: version.ID,
+	}
+
+	jobs, err := c.PredictionJobService.ListPredictionJobs(ctx, model.Project, jobQuery)
+	if err != nil {
+		log.Errorf("failed to list all prediction job for model %s version %s: %v", model.Name, version.ID, err)
+		return 500, nil, fmt.Errorf("Failed listing prediction job")
+	}
+
+	for _, item := range jobs {
+		if item.Status == models.JobPending || item.Status == models.JobRunning {
+			return 400, nil, fmt.Errorf("There are active prediction job that still using this model version")
+		}
+	}
+
+	return 200, jobs, nil
+}
+
+func (c *VersionsController) checkActiveEndpoints(ctx context.Context, model *models.Model, version *models.Version) (int, []*models.VersionEndpoint, error) {
+
+	endpoints, err := c.EndpointsService.ListEndpoints(ctx, model, version)
```

Comment:
In order to reduce the latency of this API, rather than getting all endpoints and iterating over them, it will be useful if we add some filter options to the `ListEndpoints` method, so we can directly query the DB for active endpoints. Better yet, we can have an API to only `CountEndpoints` so there is no need to fetch the actual data yet. Similarly, there can be a `DeleteEndpoints` API that accepts the `model_id` and `version_id`.

Does it make sense?

Similar comment for the predictions jobs counting. These needn't be tackled in this PR right away - they can be in a new PR.